File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/wrappers/ClipboardManager.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.wrappers;

import android.content.ClipData;
import android.content.IOnPrimaryClipChangedListener;
import android.os.Build;
import android.os.IInterface;

import java.lang.reflect.Method;

import top.saymzx.easycontrol.server.helper.FakeContext;

public class ClipboardManager {
  private static IInterface manager;
  private static Method getPrimaryClipMethod = null;
  private static Method setPrimaryClipMethod = null;
  private static Method addPrimaryClipChangedListener = null;
  private static int getMethodVersion;
  private static int setMethodVersion;
  private static int addListenerMethodVersion;

  public static void init(IInterface m) {
    manager = m;
    if (manager == null) return;
    try {
      getGetPrimaryClipMethod();
      getSetPrimaryClipMethod();
      getAddPrimaryClipChangedListenerMethod();
    } catch (Exception ignored) {
    }
  }

  private static void getGetPrimaryClipMethod() throws NoSuchMethodException {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      getPrimaryClipMethod = manager.getClass().getMethod("getPrimaryClip", String.class);
    } else {
      for (int i = 0; i < 4; i++) {
        try {
          getMethodVersion = i;
          switch (i) {
            case 0:
              getPrimaryClipMethod = manager.getClass().getMethod("getPrimaryClip", String.class, int.class);
              return;
            case 1:
              getPrimaryClipMethod = manager.getClass().getMethod("getPrimaryClip", String.class, String.class, int.class);
              return;
            case 2:
              getPrimaryClipMethod = manager.getClass().getMethod("getPrimaryClip", String.class, String.class, int.class, int.class);
              return;
            case 3:
              getPrimaryClipMethod = manager.getClass().getMethod("getPrimaryClip", String.class, int.class, String.class);
              return;
          }
        } catch (Exception ignored) {
        }
      }
    }
  }

  private static void getSetPrimaryClipMethod() throws NoSuchMethodException {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      setPrimaryClipMethod = manager.getClass().getMethod("setPrimaryClip", ClipData.class, String.class);
    } else {
      for (int i = 0; i < 3; i++) {
        try {
          setMethodVersion = i;
          switch (i) {
            case 0:
              setPrimaryClipMethod = manager.getClass().getMethod("setPrimaryClip", ClipData.class, String.class, int.class);
              return;
            case 1:
              setPrimaryClipMethod = manager.getClass().getMethod("setPrimaryClip", ClipData.class, String.class, String.class, int.class);
              return;
            case 2:
              setPrimaryClipMethod = manager.getClass().getMethod("setPrimaryClip", ClipData.class, String.class, String.class, int.class, int.class);
              return;
          }
        } catch (Exception ignored) {
        }
      }
    }
  }

  private static void getAddPrimaryClipChangedListenerMethod() throws NoSuchMethodException {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) addPrimaryClipChangedListener = manager.getClass().getMethod("addPrimaryClipChangedListener", IOnPrimaryClipChangedListener.class, String.class);
    else {
      for (int i = 0; i < 3; i++) {
        try {
          addListenerMethodVersion = i;
          switch (i) {
            case 0:
              addPrimaryClipChangedListener = manager.getClass().getMethod("addPrimaryClipChangedListener", IOnPrimaryClipChangedListener.class, String.class, int.class);
              return;
            case 1:
              addPrimaryClipChangedListener = manager.getClass().getMethod("addPrimaryClipChangedListener", IOnPrimaryClipChangedListener.class, String.class, String.class, int.class);
              return;
            case 2:
              addPrimaryClipChangedListener = manager.getClass().getMethod("addPrimaryClipChangedListener", IOnPrimaryClipChangedListener.class, String.class, String.class, int.class, int.class);
              return;
          }
        } catch (Exception ignored) {
        }
      }
    }
  }

  public static void addPrimaryClipChangedListener(IOnPrimaryClipChangedListener listener) {
    if (addPrimaryClipChangedListener == null) return;
    try {
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) addPrimaryClipChangedListener.invoke(manager, listener, FakeContext.PACKAGE_NAME);
      else {
        switch (addListenerMethodVersion) {
          case 0:
            addPrimaryClipChangedListener.invoke(manager, listener, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID);
            break;
          case 1:
            addPrimaryClipChangedListener.invoke(manager, listener, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID);
            break;
          default:
            addPrimaryClipChangedListener.invoke(manager, listener, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID, 0);
            break;
        }
      }
    } catch (Exception ignored) {
    }
  }

  public static String getText() {
    if (getPrimaryClipMethod == null) return null;
    try {
      ClipData clipData;
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) clipData = (ClipData) getPrimaryClipMethod.invoke(manager, FakeContext.PACKAGE_NAME);
      else {
        switch (getMethodVersion) {
          case 0:
            clipData = (ClipData) getPrimaryClipMethod.invoke(manager, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID);
            break;
          case 1:
            clipData = (ClipData) getPrimaryClipMethod.invoke(manager, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID);
            break;
          case 2:
            clipData = (ClipData) getPrimaryClipMethod.invoke(manager, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID, 0);
            break;
          default:
            clipData = (ClipData) getPrimaryClipMethod.invoke(manager, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID, null);
            break;
        }
      }
      if (clipData == null) return null;
      if (clipData.getItemCount() == 0) return null;
      return String.valueOf(clipData.getItemAt(0).getText());
    } catch (Exception e) {
      return null;
    }
  }

  public static void setText(String text) {
    if (setPrimaryClipMethod == null) return;
    ClipData clipData = ClipData.newPlainText("easycontrol", text);
    try {
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) setPrimaryClipMethod.invoke(manager, clipData, FakeContext.PACKAGE_NAME);
      else {
        switch (setMethodVersion) {
          case 0:
            setPrimaryClipMethod.invoke(manager, clipData, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID);
            break;
          case 1:
            setPrimaryClipMethod.invoke(manager, clipData, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID);
            break;
          default:
            setPrimaryClipMethod.invoke(manager, clipData, FakeContext.PACKAGE_NAME, null, FakeContext.ROOT_UID, 0);
            break;
        }
      }
    } catch (Exception ignored) {
    }
  }
}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/wrappers/ClipboardManager.kt
package top.saymzx.scrcpy.server.wrappers

import android.content.ClipData
import android.content.IOnPrimaryClipChangedListener
import android.os.Build
import android.os.IInterface
import top.saymzx.scrcpy.server.helper.FakeContext
import java.lang.reflect.Method

class ClipboardManager(private val manager: IInterface) {

  private var getPrimaryClipMethod: Method
  private var setPrimaryClipMethod: Method
  private var addPrimaryClipChangedListener: Method

  private var getMethodVersion = 0
  private var setMethodVersion = 0
  private var addListenerMethodVersion = 0

  val text: CharSequence?
    get() = try {
      val clipData = getPrimaryClip(getPrimaryClipMethod, getMethodVersion, manager)
      if (clipData.itemCount == 0) {
        null
      } else clipData.getItemAt(0).text
    } catch (_: Exception) {
      null
    }

  init {
    getPrimaryClipMethod = forGetPrimaryClipMethod()
    setPrimaryClipMethod = forSetPrimaryClipMethod()
    addPrimaryClipChangedListener = forAddPrimaryClipChangedListener()
  }

  private fun forGetPrimaryClipMethod(): Method {
    var field: Method
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      field = manager.javaClass.getMethod("getPrimaryClip", String::class.java)
    } else {
      try {
        field = manager.javaClass.getMethod(
          "getPrimaryClip",
          String::class.java,
          Int::class.javaPrimitiveType
        )
        getMethodVersion = 0
      } catch (e1: NoSuchMethodException) {
        try {
          field = manager.javaClass.getMethod(
            "getPrimaryClip",
            String::class.java,
            String::class.java,
            Int::class.javaPrimitiveType
          )
          getMethodVersion = 1
        } catch (e2: NoSuchMethodException) {
          try {
            field = manager.javaClass.getMethod(
              "getPrimaryClip",
              String::class.java,
              String::class.java,
              Int::class.javaPrimitiveType,
              Int::class.javaPrimitiveType
            )
            getMethodVersion = 2
          } catch (e3: NoSuchMethodException) {
            field = manager.javaClass.getMethod(
              "getPrimaryClip",
              String::class.java,
              Int::class.javaPrimitiveType,
              String::class.java
            )
            getMethodVersion = 3
          }
        }
      }
    }
    return field
  }

  private fun forSetPrimaryClipMethod(): Method {
    var field: Method
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      field =
        manager.javaClass.getMethod("setPrimaryClip", ClipData::class.java, String::class.java)
    } else {
      try {
        field = manager.javaClass.getMethod(
          "setPrimaryClip",
          ClipData::class.java,
          String::class.java,
          Int::class.javaPrimitiveType
        )
        setMethodVersion = 0
      } catch (e1: NoSuchMethodException) {
        try {
          field = manager.javaClass.getMethod(
            "setPrimaryClip",
            ClipData::class.java,
            String::class.java,
            String::class.java,
            Int::class.javaPrimitiveType
          )
          setMethodVersion = 1
        } catch (e2: NoSuchMethodException) {
          field = manager.javaClass
            .getMethod(
              "setPrimaryClip",
              ClipData::class.java,
              String::class.java,
              String::class.java,
              Int::class.javaPrimitiveType,
              Int::class.javaPrimitiveType
            )
          setMethodVersion = 2
        }
      }
    }
    return field
  }

  private fun forAddPrimaryClipChangedListener(): Method {
    var field: Method
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      field = manager.javaClass
        .getMethod(
          "addPrimaryClipChangedListener",
          IOnPrimaryClipChangedListener::class.java,
          String::class.java
        )
    } else {
      try {
        field = manager.javaClass
          .getMethod(
            "addPrimaryClipChangedListener",
            IOnPrimaryClipChangedListener::class.java,
            String::class.java,
            Int::class.javaPrimitiveType
          )
        addListenerMethodVersion = 0
      } catch (e1: NoSuchMethodException) {
        try {
          field = manager.javaClass
            .getMethod(
              "addPrimaryClipChangedListener",
              IOnPrimaryClipChangedListener::class.java,
              String::class.java,
              String::class.java,
              Int::class.javaPrimitiveType
            )
          addListenerMethodVersion = 1
        } catch (e2: NoSuchMethodException) {
          field = manager.javaClass
            .getMethod(
              "addPrimaryClipChangedListener",
              IOnPrimaryClipChangedListener::class.java,
              String::class.java,
              String::class.java,
              Int::class.javaPrimitiveType,
              Int::class.javaPrimitiveType
            )
          addListenerMethodVersion = 2
        }
      }
    }
    return field
  }

  fun setText(text: CharSequence?): Boolean {
    return try {
      val method = setPrimaryClipMethod
      val clipData = ClipData.newPlainText(null, text)
      setPrimaryClip(method, setMethodVersion, manager, clipData)
      true
    } catch (_: Exception) {
      false
    }
  }

  fun addPrimaryClipChangedListener(listener: IOnPrimaryClipChangedListener): Boolean {
    return try {
      val method = addPrimaryClipChangedListener
      addPrimaryClipChangedListener(method, addListenerMethodVersion, manager, listener)
      true
    } catch (_: Exception) {
      false
    }
  }

  companion object {
    private fun getPrimaryClip(method: Method?, methodVersion: Int, manager: IInterface): ClipData {
      return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
        method!!.invoke(manager, FakeContext.PACKAGE_NAME) as ClipData
      } else when (methodVersion) {
        0 -> method!!.invoke(manager, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID) as ClipData
        1 -> method!!.invoke(
          manager,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID
        ) as ClipData

        2 -> method!!.invoke(
          manager,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID,
          0
        ) as ClipData

        else -> method!!.invoke(
          manager,
          FakeContext.PACKAGE_NAME,
          FakeContext.ROOT_UID,
          null
        ) as ClipData
      }
    }

    private fun setPrimaryClip(
      method: Method?,
      methodVersion: Int,
      manager: IInterface,
      clipData: ClipData
    ) {
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
        method!!.invoke(manager, clipData, FakeContext.PACKAGE_NAME)
        return
      }
      when (methodVersion) {
        0 -> method!!.invoke(manager, clipData, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID)
        1 -> method!!.invoke(
          manager,
          clipData,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID
        )

        else -> method!!.invoke(
          manager,
          clipData,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID,
          0
        )
      }
    }

    private fun addPrimaryClipChangedListener(
      method: Method?,
      methodVersion: Int,
      manager: IInterface,
      listener: IOnPrimaryClipChangedListener
    ) {
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
        method!!.invoke(manager, listener, FakeContext.PACKAGE_NAME)
        return
      }
      when (methodVersion) {
        0 -> method!!.invoke(manager, listener, FakeContext.PACKAGE_NAME, FakeContext.ROOT_UID)
        1 -> method!!.invoke(
          manager,
          listener,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID
        )

        else -> method!!.invoke(
          manager,
          listener,
          FakeContext.PACKAGE_NAME,
          null,
          FakeContext.ROOT_UID,
          0
        )
      }
    }
  }
}